File: impl/php/test/producing_boundary_test.php
<?php
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
//
// NA-IETF-1 producing-boundary disclosure (the OPTIONAL, self-asserted ext key 15, §2.5.4) known-answer
// tests for the PHP SDK, graded against the independent oracle (tools/producing_boundary_oracle.py ->
// vectors/producing_boundary/cases.json), i.e. PHP == Go == Rust == Python == oracle.
//
// Five properties, mirroring impl/go/envelope/producing_boundary_test.go and
// impl/python/tests/test_producing_boundary.py, all mutation-surviving:
//   1. MATCHES ORACLE -- every case reproduces the oracle's body-no-id / content-id / full-body bytes;
//      a real ML-DSA-65 sign+verify yields exactly the oracle's accept/reject verdict; and the parsed
//      disclosure (present/kind/boundary/reporting) matches. Non-canonical sub-map bytes are rejected
//      NonCanonical at the codec.
//   2. UNDER SIGNATURE -- the disclosure is folded into the SIGNER's signed body: splicing a different
//      boundary into a signed object (keeping its id + signature) is rejected.
//   3. READER ROUND-TRIP -- set/get carries the value; the field is OPTIONAL; the setter DROPS a
//      reporting-boundary under observed (an observer relays from no one).
//   4. MALFORMED IGNORED [MUTATION ANCHOR] -- a well-formed object carrying a MALFORMED disclosure
//      (reporting under observed) in the non-critical ext map still verifies and is NOT surfaced.
//   5. CEXT REJECTED [MUTATION ANCHOR] -- the disclosure in the CRITICAL cext map is UnknownCriticalExt.
//
// Run: php -d extension=sodium -d extension=intl -d ffi.enable=1 -d extension=ffi
//          test/producing_boundary_test.php   (from impl/php/). Exit code 0 = all checks passed.
require __DIR__ . '/../src/bootstrap.php';

use Naalp\Cbor;
use Naalp\Cose;
use Naalp\Envelope;
use Naalp\MlDsa;
use Naalp\NaalpObject;
use Naalp\ProducingBoundary;
use Naalp\U;
use Naalp\B;
use Naalp\T;
use Naalp\M;

// A LOCAL test signing seed -- the verdict is a sign+verify round-trip, not a reproduction of the
// oracle's signature (full_hex is the object BODY, not a signed COSE object, so byte-parity needs no
// signing).
$SEED = pack("C*", ...range(0, 31));

$fails = 0;
function check(string $name, string $got, string $want): void
{
    global $fails;
    if ($got === $want) {
        echo "  ok   $name\n";
    } else {
        $fails++;
        echo "  FAIL $name\n       got  $got\n       want $want\n";
    }
}

/** Assert that $fn throws a Throwable whose ->kind === $wantKind. */
function expect_kind(string $name, callable $fn, string $wantKind): void
{
    global $fails;
    try {
        $fn();
        $fails++;
        echo "  FAIL $name (no exception; expected $wantKind)\n";
    } catch (\Throwable $e) {
        $got = $e->kind ?? get_class($e);
        if ($got === $wantKind) {
            echo "  ok   $name\n";
        } else {
            $fails++;
            echo "  FAIL $name\n       got kind  $got\n       want kind $wantKind\n";
        }
    }
}

/** Assert that $fn throws SOME Throwable (kind unspecified -- a splice is rejected at whichever check
 * catches it first, mirroring the generic assertRaises(Exception)/verr==nil style the Python and Go
 * worked examples use for this property). */
function expect_throws(string $name, callable $fn): void
{
    global $fails;
    try {
        $fn();
        $fails++;
        echo "  FAIL $name (no exception thrown)\n";
    } catch (\Throwable $e) {
        $got = $e->kind ?? get_class($e);
        echo "  ok   $name ($got)\n";
    }
}

function kindOk(int $ch, int $k): bool
{
    return true;
}

function findPbVector(): ?string
{
    $d = __DIR__;
    for ($i = 0; $i < 6; $i++) {
        $p = $d . '/vectors/producing_boundary/cases.json';
        if (is_file($p)) {
            return $p;
        }
        $d = dirname($d);
    }
    return null;
}

/** Build the shared base object (fields 2..10) from the corpus's LOGICAL fields -- never from the
 * oracle hex, so a constant encoder diverges from the pinned bytes. */
function baseObject(array $corpus): NaalpObject
{
    $base = $corpus["base_object"];
    $causes = [];
    foreach ($base["causes_hex"] as $h) {
        $causes[] = hex2bin($h);
    }
    return new NaalpObject(
        kind: $base["kind"],
        channel: $base["channel"],
        signer: hex2bin($base["signer_hex"]),
        created: $base["created"],
        effect: $base["effect"],
        body: new T($base["body_str"]),
        tier: $base["tier"],
        profile: $base["profile"],
        causes: $causes,
    );
}

/** Build the ext[15]/cext[15] sub-map DIRECTLY from the case's logical fields (the ONLY variable per
 * case) using the WIRE sub-keys 1/2/3, reproducing the oracle bytes for well-formed AND malformed
 * values -- the malformed cases cannot be built via Envelope::setProducingBoundary by design, so they
 * are constructed here. */
function applyPlacement(NaalpObject $o, array $tc): void
{
    if ($tc["placement"] === "absent") {
        return;
    }
    $sub = [];
    if (($tc["boundary_hex"] ?? null) !== null) {
        $sub[] = [new U(1), new B(hex2bin($tc["boundary_hex"]))];
    }
    if (($tc["kind"] ?? null) !== null) {
        $sub[] = [new U(2), new U($tc["kind"])];
    }
    if (($tc["reporting_hex"] ?? null) !== null) {
        $sub[] = [new U(3), new B(hex2bin($tc["reporting_hex"]))];
    }
    $ext = new M([[new U(Envelope::PRODUCING_BOUNDARY_KEY), new M($sub)]]);
    if ($tc["placement"] === "ext") {
        $o->ext = $ext;
    } elseif ($tc["placement"] === "cext") {
        $o->cext = $ext;
    } else {
        throw new \RuntimeException("unknown placement " . $tc["placement"]);
    }
}

/** Full end-to-end pure-ML-DSA-65 sign of an object (build -> protected header -> sign -> assemble),
 * mirroring the tag-18 pattern rotation_test.php uses inline (there is no generic Envelope::sign() for
 * a pure object; only signWithEd25519/signComposite/signRotationObject exist). */
function signPure(NaalpObject $o, string $seed): string
{
    $payload = Envelope::buildPayload($o); // sets id
    $prot = Envelope::protectedHeader(Cose::ALG_MLDSA65, $o->signer, $o->profile);
    $tbs = Cose::toBeSignedRaw($prot, $payload);
    $sig = MlDsa::sign($seed, $tbs, Cose::ALG_MLDSA65);
    return Cose::assembleSign1Raw($prot, $payload, $sig);
}

/** Sign a RAW payload (used for the negatives: a non-canonical sub-map inside an otherwise
 * well-formed body, built by hand from the oracle's payload_hex, not via buildPayload). */
function signRawPayload(NaalpObject $o, string $seed, string $payload): string
{
    $prot = Envelope::protectedHeader(Cose::ALG_MLDSA65, $o->signer, $o->profile);
    $tbs = Cose::toBeSignedRaw($prot, $payload);
    $sig = MlDsa::sign($seed, $tbs, Cose::ALG_MLDSA65);
    return Cose::assembleSign1Raw($prot, $payload, $sig);
}

echo "NA-IETF-1 producing-boundary disclosure KAT (PHP)\n";

if (!MlDsa::available()) {
    echo "  SKIP all producing-boundary checks: ML-DSA unavailable (" . MlDsa::unavailableReason() . ")\n";
    echo "PASS (skipped)\n";
    exit(0);
}

$vecPath = findPbVector();
if ($vecPath === null) {
    echo "  SKIP all producing-boundary checks: committed vector not present (standalone install)\n";
    echo "PASS (skipped)\n";
    exit(0);
}
$corpus = json_decode(file_get_contents($vecPath), true);
check("producing_boundary_key matches", (string) $corpus["producing_boundary_key"], (string) Envelope::PRODUCING_BOUNDARY_KEY);

$pk = MlDsa::keygenFromSeed($SEED, Cose::ALG_MLDSA65);

// 1. MATCHES ORACLE.
echo "\n1. MatchesOracle\n";
foreach ($corpus["cases"] as $tc) {
    $name = $tc["name"];

    // byte parity: body-without-id, content id, full body (all pre-signature).
    $o = baseObject($corpus);
    applyPlacement($o, $tc);
    check("$name body-no-id", bin2hex(Cbor::encode($o->bodyMap(false))), $tc["body_no_id_hex"]);
    $cid = $o->contentId();
    check("$name content-id", bin2hex($cid), $tc["content_id_hex"]);
    $o->id = $cid;
    check("$name full-body", bin2hex(Cbor::encode($o->bodyMap(true))), $tc["full_hex"]);

    // verdict: sign for real + verify offline; assert accept vs the named error.
    $o2 = baseObject($corpus);
    applyPlacement($o2, $tc);
    $signed = signPure($o2, $SEED);
    if ($tc["expect"] === "accept") {
        $got = Envelope::verify(Cose::PROFILE_PUBLIC, Cose::ALG_MLDSA65, $pk, 'kindOk', $signed);
        [$pb, $present] = Envelope::producingBoundary($got);
        check("$name present", $present ? "true" : "false", $tc["present"] ? "true" : "false");
        if ($present && $tc["surfaced"] !== null) {
            $s = $tc["surfaced"];
            check("$name kind", (string) $pb->kind, (string) $s["kind"]);
            check("$name boundary", bin2hex($pb->boundary), $s["boundary_hex"]);
            $wantRep = $s["reporting_hex"] ?? "";
            check("$name reporting", bin2hex($pb->reporting ?? ""), $wantRep);
        }
    } else {
        expect_kind("$name verdict", function () use ($pk, $signed) {
            Envelope::verify(Cose::PROFILE_PUBLIC, Cose::ALG_MLDSA65, $pk, 'kindOk', $signed);
        }, $tc["expect"]);
    }
}

// non-canonical disclosure bodies (sub-map keys out of order) are rejected at the CBOR layer.
foreach ($corpus["negatives"] ?? [] as $neg) {
    $o = baseObject($corpus);
    $payload = hex2bin($neg["payload_hex"]);
    $signed = signRawPayload($o, $SEED, $payload);
    expect_kind("negative_" . $neg["name"], function () use ($pk, $signed) {
        Envelope::verify(Cose::PROFILE_PUBLIC, Cose::ALG_MLDSA65, $pk, 'kindOk', $signed);
    }, $neg["expect"]);
}

// 2. UNDER SIGNATURE.
echo "\n2. UnderSignature\n";
$o = baseObject($corpus);
Envelope::setProducingBoundary(
    $o,
    new ProducingBoundary(hex2bin("424f554e444152595f58"), Envelope::PRODUCING_BOUNDARY_OBSERVED)
);
$signed = signPure($o, $SEED);
$got = Envelope::verify(Cose::PROFILE_PUBLIC, Cose::ALG_MLDSA65, $pk, 'kindOk', $signed);
[$pb, $present] = Envelope::producingBoundary($got);
check("under-signature read-back", ($present && $pb->kind === Envelope::PRODUCING_BOUNDARY_OBSERVED) ? "ok" : "fail", "ok");

// tamper: change boundary, keep the original id, reuse the original signature (a splice).
$tampered = baseObject($corpus);
Envelope::setProducingBoundary(
    $tampered,
    new ProducingBoundary(hex2bin("4f524947494e5f59"), Envelope::PRODUCING_BOUNDARY_OBSERVED)
);
$tampered->id = $o->id; // keep original content id -- a splice, not a re-sign
$payload = Cbor::encode($tampered->bodyMap(true));
[$prot, , $sig] = Cose::parseSign1Raw($signed);
$forged = Cose::assembleSign1Raw($prot, $payload, $sig);
expect_throws("under-signature tamper rejected", function () use ($pk, $forged) {
    Envelope::verify(Cose::PROFILE_PUBLIC, Cose::ALG_MLDSA65, $pk, 'kindOk', $forged);
});

// 3. READER ROUND-TRIP.
echo "\n3. ReaderRoundTrip\n";
$o = baseObject($corpus);
[$_, $present] = Envelope::producingBoundary($o);
check("fresh object has no disclosure", $present ? "present" : "absent", "absent");
$x = hex2bin("424f554e444152595f58");
$y = hex2bin("4f524947494e5f59");

Envelope::setProducingBoundary($o, new ProducingBoundary($x, Envelope::PRODUCING_BOUNDARY_REPORTED, $y));
[$pb, $present] = Envelope::producingBoundary($o);
check("reported round-trip present", $present ? "true" : "false", "true");
check("reported round-trip kind", (string) $pb->kind, (string) Envelope::PRODUCING_BOUNDARY_REPORTED);
check("reported round-trip boundary", bin2hex($pb->boundary), bin2hex($x));
check("reported round-trip reporting", bin2hex($pb->reporting), bin2hex($y));

// the setter drops a reporting-boundary under observed: read-back has no reporting.
Envelope::setProducingBoundary($o, new ProducingBoundary($x, Envelope::PRODUCING_BOUNDARY_OBSERVED, $y));
[$pb, $present] = Envelope::producingBoundary($o);
check("observed round-trip present", $present ? "true" : "false", "true");
check("observed round-trip kind", (string) $pb->kind, (string) Envelope::PRODUCING_BOUNDARY_OBSERVED);
check("setter drops reporting under observed", $pb->reporting === null ? "null" : "notnull", "null");

// 4. MALFORMED IGNORED [MUTATION ANCHOR]: dropping the reporting-under-observed check in
// Envelope::producingBoundary() flips present false->true and this check ok->FAIL; that check is the
// observer-relays-from-no-one invariant.
echo "\n4. MalformedIgnored [MUTATION ANCHOR]\n";
$o = baseObject($corpus);
// malformed ext[15] = {1:X, 2:observed, 3:Y} built directly (the setter refuses to build it).
$o->ext = new M([[new U(Envelope::PRODUCING_BOUNDARY_KEY), new M([
    [new U(1), new B(hex2bin("424f554e444152595f58"))],
    [new U(2), new U(Envelope::PRODUCING_BOUNDARY_OBSERVED)],
    [new U(3), new B(hex2bin("4f524947494e5f59"))],
])]]);
$signed = signPure($o, $SEED);
$got = Envelope::verify(Cose::PROFILE_PUBLIC, Cose::ALG_MLDSA65, $pk, 'kindOk', $signed); // must NOT throw (may-ignore)
echo "  ok   malformed disclosure still verifies (no exception)\n";
[$_, $present] = Envelope::producingBoundary($got);
check("malformed disclosure not surfaced", $present ? "present" : "absent", "absent");

// 5. CEXT REJECTED [MUTATION ANCHOR]: the disclosure in the CRITICAL cext map (field 12) is an
// unrecognized critical extension -> UnknownCriticalExt, fail-closed. A disclosure must never
// masquerade as a must-understand gate.
echo "\n5. CextRejected [MUTATION ANCHOR]\n";
$o = baseObject($corpus);
$o->cext = new M([[new U(Envelope::PRODUCING_BOUNDARY_KEY), new M([
    [new U(1), new B(hex2bin("424f554e444152595f58"))],
    [new U(2), new U(Envelope::PRODUCING_BOUNDARY_OBSERVED)],
])]]);
$signed = signPure($o, $SEED);
expect_kind("cext producing-boundary rejected", function () use ($pk, $signed) {
    Envelope::verify(Cose::PROFILE_PUBLIC, Cose::ALG_MLDSA65, $pk, 'kindOk', $signed);
}, "UnknownCriticalExt");

echo "\n" . ($fails === 0 ? "PASS" : "FAIL ($fails)") . "\n";
exit($fails === 0 ? 0 : 1);

File: impl/go/envelope/producing_boundary_test.go
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.

package envelope

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/bubblefish-tech/naalp_protocol/impl/go/cbor"
	"github.com/bubblefish-tech/naalp_protocol/impl/go/cose"
)

// --- NA-IETF-1 producing-boundary disclosure (the OPTIONAL, self-asserted ext key 15, §2.5.4) ----

const producingBoundaryVectorPath = "../../../vectors/producing_boundary/cases.json"

type pbSurfaced struct {
	Kind         uint64  `json:"kind"`
	BoundaryHex  string  `json:"boundary_hex"`
	ReportingHex *string `json:"reporting_hex"`
}

type pbCase struct {
	Name         string      `json:"name"`
	Placement    string      `json:"placement"`
	BoundaryHex  *string     `json:"boundary_hex"`
	Kind         *uint64     `json:"kind"`
	ReportingHex *string     `json:"reporting_hex"`
	Present      bool        `json:"present"`
	Surfaced     *pbSurfaced `json:"surfaced"`
	BodyNoIDHex  string      `json:"body_no_id_hex"`
	ContentIDHex string      `json:"content_id_hex"`
	FullHex      string      `json:"full_hex"`
	Expect       string      `json:"expect"`
}

type pbCorpus struct {
	Key  uint64 `json:"producing_boundary_key"`
	Base struct {
		Kind      uint64   `json:"kind"`
		Channel   uint64   `json:"channel"`
		Tier      uint64   `json:"tier"`
		SignerHex string   `json:"signer_hex"`
		Created   uint64   `json:"created"`
		Effect    uint64   `json:"effect"`
		CausesHex []string `json:"causes_hex"`
		Profile   uint64   `json:"profile"`
		BodyStr   string   `json:"body_str"`
	} `json:"base_object"`
	Cases     []pbCase `json:"cases"`
	Negatives []struct {
		Name       string `json:"name"`
		PayloadHex string `json:"payload_hex"`
		Expect     string `json:"expect"`
	} `json:"negatives"`
}

func loadPB(t *testing.T) pbCorpus {
	t.Helper()
	b, err := os.ReadFile(filepath.Clean(producingBoundaryVectorPath))
	if err != nil {
		t.Fatalf("read producing_boundary corpus: %v", err)
	}
	var c pbCorpus
	if err := json.Unmarshal(b, &c); err != nil {
		t.Fatalf("parse producing_boundary corpus: %v", err)
	}
	return c
}

// basePBObject builds the shared base object (fields 2..10) from logical fields — never from the
// oracle hex, so a constant encoder diverges from the pinned bytes.
func basePBObject(t *testing.T, corpus pbCorpus) *Object {
	t.Helper()
	var causes [][]byte
	for _, h := range corpus.Base.CausesHex {
		causes = append(causes, mh(t, h))
	}
	return &Object{
		Kind: corpus.Base.Kind, Channel: corpus.Base.Channel, Tier: corpus.Base.Tier,
		Signer: mh(t, corpus.Base.SignerHex), Created: corpus.Base.Created, Effect: corpus.Base.Effect,
		Causes: causes, Profile: corpus.Base.Profile, Body: cbor.Tstr(corpus.Base.BodyStr),
	}
}

// applyPBPlacement builds the ext[15]/cext[15] sub-map DIRECTLY from the case's logical fields — the
// ONLY variable per case — reproducing the oracle bytes for well-formed AND malformed values (the
// malformed cases cannot be built via SetProducingBoundary by design, so they are constructed here).
func applyPBPlacement(t *testing.T, o *Object, tc pbCase) {
	t.Helper()
	if tc.Placement == "absent" {
		return
	}
	var sub cbor.Map
	if tc.BoundaryHex != nil {
		sub = append(sub, cbor.Pair{K: cbor.Uint(pbFieldBoundary), V: cbor.Bstr(mh(t, *tc.BoundaryHex))})
	}
	if tc.Kind != nil {
		sub = append(sub, cbor.Pair{K: cbor.Uint(pbFieldKind), V: cbor.Uint(*tc.Kind)})
	}
	if tc.ReportingHex != nil {
		sub = append(sub, cbor.Pair{K: cbor.Uint(pbFieldReporting), V: cbor.Bstr(mh(t, *tc.ReportingHex))})
	}
	ext := cbor.Map{{K: cbor.Uint(ProducingBoundaryKey), V: sub}}
	switch tc.Placement {
	case "ext":
		o.Ext = ext
	case "cext":
		o.Cext = ext
	default:
		t.Fatalf("unknown placement %q", tc.Placement)
	}
}

// TestProducingBoundaryMatchesOracle grades disclosure body bytes, accept/reject verdicts, AND the
// parsed disclosure (present/kind/boundary/reporting) against the independent oracle
// (tools/producing_boundary_oracle.py): every body_no_id_hex/content_id_hex/full_hex is
// byte-identical, and every case verifies or fails with exactly the oracle's verdict over a REAL
// ML-DSA-65 signed object. knownCext is nil throughout — the disclosure is enforced by the
// envelope's existing rules, not the caller's critical-extension set.
func TestProducingBoundaryMatchesOracle(t *testing.T) {
	corpus := loadPB(t)
	if corpus.Key != ProducingBoundaryKey {
		t.Fatalf("producing_boundary_key: corpus %d != impl %d", corpus.Key, ProducingBoundaryKey)
	}
	s, v := testSigner(t)
	for _, tc := range corpus.Cases {
		tc := tc
		t.Run(tc.Name, func(t *testing.T) {
			o := basePBObject(t, corpus)
			applyPBPlacement(t, o, tc)

			// byte parity: body-without-id, content id, full body.
			bn, _ := cbor.Encode(o.bodyMap(false))
			if got := hex.EncodeToString(bn); got != tc.BodyNoIDHex {
				t.Fatalf("body-no-id\n got %s\nwant %s", got, tc.BodyNoIDHex)
			}
			id, err := o.ContentID()
			if err != nil {
				t.Fatal(err)
			}
			if got := hex.EncodeToString(id); got != tc.ContentIDHex {
				t.Fatalf("content-id\n got %s\nwant %s", got, tc.ContentIDHex)
			}
			o.ID = id
			full, _ := cbor.Encode(o.bodyMap(true))
			if got := hex.EncodeToString(full); got != tc.FullHex {
				t.Fatalf("full-body\n got %s\nwant %s", got, tc.FullHex)
			}

			// verdict: sign for real and verify offline; assert accept vs the named error.
			o2 := basePBObject(t, corpus)
			applyPBPlacement(t, o2, tc)
			signed, err := Sign(o2, s)
			if err != nil {
				t.Fatalf("sign: %v", err)
			}
			got, verr := Verify(cose.ProfilePublic, v, acceptKind, nil, signed)
			if tc.Expect == "accept" {
				if verr != nil {
					t.Fatalf("expected accept, got %v", verr)
				}
				pb, present := got.ProducingBoundary()
				if present != tc.Present {
					t.Fatalf("ProducingBoundary present: got %v want %v", present, tc.Present)
				}
				if present {
					if tc.Surfaced == nil {
						t.Fatal("corpus marks present but carries no surfaced disclosure")
					}
					if pb.Kind != tc.Surfaced.Kind {
						t.Fatalf("kind: got %d want %d", pb.Kind, tc.Surfaced.Kind)
					}
					if got := hex.EncodeToString(pb.Boundary); got != tc.Surfaced.BoundaryHex {
						t.Fatalf("boundary: got %s want %s", got, tc.Surfaced.BoundaryHex)
					}
					wantReporting := ""
					if tc.Surfaced.ReportingHex != nil {
						wantReporting = *tc.Surfaced.ReportingHex
					}
					if got := hex.EncodeToString(pb.Reporting); got != wantReporting {
						t.Fatalf("reporting: got %s want %s", got, wantReporting)
					}
				}
			} else {
				ce, ok := verr.(*cose.Error)
				if !ok || ce.Kind != tc.Expect {
					t.Fatalf("expected %s, got %v", tc.Expect, verr)
				}
			}
		})
	}

	// non-canonical disclosure bodies (sub-map keys out of order) are rejected at the CBOR layer.
	for _, neg := range corpus.Negatives {
		neg := neg
		t.Run("negative_"+neg.Name, func(t *testing.T) {
			obj := signRawPayload(t, s, mh(t, neg.PayloadHex))
			_, verr := Verify(cose.ProfilePublic, v, acceptKind, nil, obj)
			ce, ok := verr.(*cose.Error)
			if !ok || ce.Kind != neg.Expect {
				t.Fatalf("expected %s, got %v", neg.Expect, verr)
			}
		})
	}
}

// TestProducingBoundaryUnderSignature proves the disclosure is folded into the SIGNER's COSE_Sign1
// signed input (ext, field 11, is part of the signed body): changing the boundary in a signed
// object's payload without re-signing breaks verification. A self-asserted disclosure that is NOT
// under the signer's signature would be forgeable, defeating attributability.
func TestProducingBoundaryUnderSignature(t *testing.T) {
	corpus := loadPB(t)
	s, v := testSigner(t)
	o := basePBObject(t, corpus)
	o.SetProducingBoundary(ProducingBoundary{Boundary: mh(t, "424f554e444152595f58"), Kind: ProducingBoundaryObserved})
	signed, err := Sign(o, s)
	if err != nil {
		t.Fatalf("sign: %v", err)
	}
	// baseline: the signed object verifies and reads back the disclosure.
	got, err := Verify(cose.ProfilePublic, v, acceptKind, nil, signed)
	if err != nil {
		t.Fatalf("verify valid: %v", err)
	}
	if pb, present := got.ProducingBoundary(); !present || pb.Kind != ProducingBoundaryObserved {
		t.Fatalf("disclosure read-back: got (%+v,%v)", pb, present)
	}
	// tamper: change the boundary and re-encode the body WITHOUT re-signing; keep the original id
	// and reuse the original signature bytes — a real forgery attempt that MUST be rejected.
	tampered := basePBObject(t, corpus)
	tampered.SetProducingBoundary(ProducingBoundary{Boundary: mh(t, "4f524947494e5f59"), Kind: ProducingBoundaryObserved})
	tampered.ID = o.ID // keep the original content id — a splice, not a re-sign
	payload, _ := cbor.Encode(tampered.bodyMap(true))
	prot, _ := protectedHeader(s.Alg(), tampered.Signer, tampered.Profile)
	_, _, origSig, _ := cose.ParseSign1Raw(signed)
	forged, _ := cose.AssembleSign1Raw(prot, payload, origSig)
	if _, verr := Verify(cose.ProfilePublic, v, acceptKind, nil, forged); verr == nil {
		t.Fatal("tampered producing-boundary must be rejected (it is under signature), got accept")
	}
}

// TestProducingBoundaryReaderRoundTrip proves SetProducingBoundary/ProducingBoundary carry the value,
// that the field is OPTIONAL (a fresh object has none), that a reported disclosure carries its
// reporting-boundary, and that the setter DROPS a reporting-boundary under observed (an observer
// relays from no one) so a caller cannot accidentally build a malformed disclosure.
func TestProducingBoundaryReaderRoundTrip(t *testing.T) {
	corpus := loadPB(t)
	o := basePBObject(t, corpus)
	if _, present := o.ProducingBoundary(); present {
		t.Fatal("fresh object must have no producing-boundary disclosure")
	}
	x := mh(t, "424f554e444152595f58")
	y := mh(t, "4f524947494e5f59")

	o.SetProducingBoundary(ProducingBoundary{Boundary: x, Kind: ProducingBoundaryReported, Reporting: y})
	pb, present := o.ProducingBoundary()
	if !present || pb.Kind != ProducingBoundaryReported ||
		hex.EncodeToString(pb.Boundary) != "424f554e444152595f58" ||
		hex.EncodeToString(pb.Reporting) != "4f524947494e5f59" {
		t.Fatalf("reported disclosure round-trip wrong: %+v present=%v", pb, present)
	}

	// the setter drops a reporting-boundary under observed: the read-back must have no reporting.
	o.SetProducingBoundary(ProducingBoundary{Boundary: x, Kind: ProducingBoundaryObserved, Reporting: y})
	pb, present = o.ProducingBoundary()
	if !present || pb.Kind != ProducingBoundaryObserved || pb.Reporting != nil {
		t.Fatalf("observed disclosure must drop reporting: %+v present=%v", pb, present)
	}
}

// TestProducingBoundaryMalformedIgnored is the MUTATION ANCHOR for the may-ignore rule: a well-formed
// object carrying a MALFORMED producing-boundary in the non-critical ext map (a reporting-boundary
// under observed) still Signs and Verifies, and the disclosure is NOT surfaced. Removing the
// "reporting under observed -> malformed" check in ProducingBoundary flips present false->true and
// this test pass->fail; that check is the observer-relays-from-no-one invariant.
func TestProducingBoundaryMalformedIgnored(t *testing.T) {
	corpus := loadPB(t)
	s, v := testSigner(t)
	o := basePBObject(t, corpus)
	// build the malformed ext[15] = {1:X, 2:observed, 3:Y} directly (the setter refuses to build it).
	o.Ext = cbor.Map{{K: cbor.Uint(ProducingBoundaryKey), V: cbor.Map{
		{K: cbor.Uint(pbFieldBoundary), V: cbor.Bstr(mh(t, "424f554e444152595f58"))},
		{K: cbor.Uint(pbFieldKind), V: cbor.Uint(ProducingBoundaryObserved)},
		{K: cbor.Uint(pbFieldReporting), V: cbor.Bstr(mh(t, "4f524947494e5f59"))},
	}}}
	signed, err := Sign(o, s)
	if err != nil {
		t.Fatalf("sign: %v", err)
	}
	got, err := Verify(cose.ProfilePublic, v, acceptKind, nil, signed)
	if err != nil {
		t.Fatalf("a malformed non-critical disclosure must be ignored, not rejected (may-ignore), got %v", err)
	}
	if _, present := got.ProducingBoundary(); present {
		t.Fatal("a malformed disclosure (reporting under observed) must NOT be surfaced")
	}
}

// TestProducingBoundaryCextRejected is the MUTATION ANCHOR for the fail-closed rule: the disclosure
// placed in the CRITICAL cext map (field 12) is an unrecognized critical extension and the object is
// rejected UnknownCriticalExt. A disclosure must never masquerade as a must-understand gate.
func TestProducingBoundaryCextRejected(t *testing.T) {
	corpus := loadPB(t)
	s, v := testSigner(t)
	o := basePBObject(t, corpus)
	o.Cext = cbor.Map{{K: cbor.Uint(ProducingBoundaryKey), V: cbor.Map{
		{K: cbor.Uint(pbFieldBoundary), V: cbor.Bstr(mh(t, "424f554e444152595f58"))},
		{K: cbor.Uint(pbFieldKind), V: cbor.Uint(ProducingBoundaryObserved)},
	}}}
	signed, err := Sign(o, s)
	if err != nil {
		t.Fatalf("sign: %v", err)
	}
	_, verr := Verify(cose.ProfilePublic, v, acceptKind, nil, signed)
	ce, ok := verr.(*cose.Error)
	if !ok || ce.Kind != "UnknownCriticalExt" {
		t.Fatalf("cext producing-boundary must be rejected UnknownCriticalExt, got %v", verr)
	}
}
